src/Microsoft.IdentityModel.Xml/EncryptionMethod.cs

```diff
@@ -0,0 +1,122 @@
+    /// Represents the <see cref="EncryptionMethod"/> element in XML encryption that describes the encryption algorithm applied to the cipher data. This class cannot be inherited.
+    /// </summary>
+    /// <remarks> http://www.w3.org/TR/xmlenc-core/#sec-EncryptionMethod </remarks>
+    public class EncryptionMethod
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EncryptionMethod"/> class.
+        /// </summary>
+        public EncryptionMethod()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EncryptionMethod"/> class specifying an algorithm Uniform Resource Identifier(URI).
+        /// </summary>
+        /// <param name="algorithm"></param>
+        public EncryptionMethod(string algorithm)
+        {
+            KeyAlgorithm = algorithm;
+        }
+
+        /// <summary>
+        /// Gets or sets a Uniform Resource Identifier (URI) that describes the algorithm to use for XML encryption.
+        /// </summary>
+        public string KeyAlgorithm { get; set; }
+
+        /// <summary>
+        /// Gets or sets the digest method Uniform Resource Identifier (URI)
+        /// </summary>
+        public string DigestMethod { get; set; }
+
+        internal void WriteXml(XmlWriter writer)
+        {
+            if (writer == null)
+                throw LogArgumentNullException(nameof(writer));
+
+            // nothing to write - return
+            if (string.IsNullOrEmpty(KeyAlgorithm))
+                return;
+
+            writer.WriteStartElement(XmlEncryptionConstants.Prefix, XmlEncryptionConstants.Elements.EncryptionMethod, XmlEncryptionConstants.Namespace);
+            writer.WriteAttributeString(XmlEncryptionConstants.Attributes.Algorithm, null, KeyAlgorithm);
+
+            if (!string.IsNullOrEmpty(DigestMethod))
+            {
+                writer.WriteStartElement(XmlSignatureConstants.PreferredPrefix, XmlSignatureConstants.Elements.DigestMethod, XmlSignatureConstants.Namespace);
+                writer.WriteAttributeString(XmlSignatureConstants.Attributes.Algorithm, null, DigestMethod);
+                writer.WriteEndElement();
+            }
+
+            writer.WriteEndElement();
+        }
+
+        internal void ReadXml(XmlDictionaryReader reader)
+        {
+            if (reader == null)
+                throw LogArgumentNullException(nameof(reader));
+
+            if (reader.IsStartElement(XmlEncryptionConstants.Elements.EncryptionMethod, XmlEncryptionConstants.Namespace))
+            {
+                KeyAlgorithm = reader.GetAttribute(XmlEncryptionConstants.Attributes.Algorithm, null);
+
+                if (reader.IsEmptyElement)
+                {
+                    reader.Read();
+                    return;
+                }
+
+                reader.ReadStartElement(XmlEncryptionConstants.Elements.EncryptionMethod, XmlEncryptionConstants.Namespace);
+
+                while (reader.IsStartElement())
```

Comment:
>while (reader.IsStartElement()) [](start = 16, length = 31)

what is reader.IsSTartElement supposed to do? Reading this code I would be afraid that it would either potentially just have one element to read, however if it never reads anything this will be an endless loop

Reply:
IsStartElement checks if xml reader is pointing to a start tag or empty element tag.
In case where there is nothing to read, reader will be pointing to an EndElement and exit from the while loop.

This logic is implemented because EncryptionMethod supports any element as a child. For now, we care only about DigestMethod and we are ignoring (but logging) other elements that may appear.

That said, it might be better to use reader.Skip() instead of reader.Read() on ln. 109 as DigestMethod by xenc xsd schema might have child elements (it's usually an empty element)